### PR TITLE
Sync working directory with all processes when using Hydra

### LINKED ISSUE #5512: FileNotFoundError for best checkpoint when using DDP with Hydra
## 🐛 Bug

I am getting a FileNotFoundError for loading the best checkpoint when using `trainer.test()` after `trainer.fit()` in DDP mode with [Hydra](https://hydra.cc/).

My configuration file specifies that `hydra.run.dir="/path/to/data/${now:%Y-%m-%d_%H-%M-%S}"`.
As a result, the first process (rank 0) spawns in "/path/to/data/datetime1" and creates the "ckpts" and "logs" folders there while the second process (rank 1) spawns in "/path/to/data/datetime2" and cannot access the "ckpts" and "logs" folders.
It appears that when calling `trainer.test()`, the program looks for "/path/to/data/datetime2/ckpts/best.ckpt" which is indeed not there.

Here is the error stack:
```
Epoch 4: val_acc reached 30.00000 (best 30.00000), saving model to /home/azouaoui/github/PL-Hydra-template/data/runs/2021-01-14_08-03-33/ckpts/epoch=004-val_acc=30.000.ckpt as top 1
[lightning][INFO] - Epoch 4: val_acc reached 30.00000 (best 30.00000), saving model to /home/azouaoui/github/PL-Hydra-template/data/runs/2021-01-14_08-03-33/ckpts/epoch=004-val_acc=30.000.ckpt as top 1
Epoch 4: 100%|███████████| 4/4 [00:00<00:00, 11.70it/s, loss=2.729, v_num=0, val_acc=30, best_val_acc=30Saving latest checkpoint...                                                                               
[lightning][INFO] - Saving latest checkpoint...
Epoch 4: 100%|███████████| 4/4 [00:00<00:00, 11.62it/s, loss=2.729, v_num=0, val_acc=30, best_val_acc=30]
[__main__][CRITICAL] - [Errno 2] No such file or directory: '/home/azouaoui/github/PL-Hydra-template/data/runs/2021-01-14_08-03-35/ckpts/epoch=004-val_acc=30.000.ckpt'
Traceback (most recent call last):
  File "/home/azouaoui/github/PL-Hydra-template/train.py", line 41, in main
    train(cfg)
  File "/home/azouaoui/github/PL-Hydra-template/train.py", line 34, in train
    logger.info(trainer.test())
  File "/scratch/artemis/azouaoui/miniconda3/envs/jz/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 720, in test
    results = self.__test_using_best_weights(ckpt_path, test_dataloaders)
  File "/scratch/artemis/azouaoui/miniconda3/envs/jz/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 750, in __test_using_best_weights
    ckpt = pl_load(ckpt_path, map_location=lambda storage, loc: storage)
  File "/scratch/artemis/azouaoui/miniconda3/envs/jz/lib/python3.7/site-packages/pytorch_lightning/utilities/cloud_io.py", line 31, in load
    with fs.open(path_or_url, "rb") as f:
  File "/scratch/artemis/azouaoui/miniconda3/envs/jz/lib/python3.7/site-packages/fsspec/spec.py", line 936, in open
    **kwargs
  File "/scratch/artemis/azouaoui/miniconda3/envs/jz/lib/python3.7/site-packages/fsspec/implementations/local.py", line 117, in _open
    return LocalFileOpener(path, mode, fs=self, **kwargs)
  File "/scratch/artemis/azouaoui/miniconda3/envs/jz/lib/python3.7/site-packages/fsspec/implementations/local.py", line 199, in __init__
    self._open()
  File "/scratch/artemis/azouaoui/miniconda3/envs/jz/lib/python3.7/site-packages/fsspec/implementations/local.py", line 204, in _open
    self.f = open(self.path, mode=self.mode)
FileNotFoundError: [Errno 2] No such file or directory: '/home/azouaoui/github/PL-Hydra-template/data/runs/2021-01-14_08-03-35/ckpts/epoch=004-val_acc=30.000.ckpt'
```

## Please reproduce using the BoringModel

Error is triggered by using DDP with at least 2 GPUs. Hence I cannot use Colab.

### To Reproduce

Use this [repository](https://github.com/inzouzouwetrust/PL-Hydra-DDP-bug)

Have at least 2 GPUs available.
```
$ git clone https://github.com/inzouzouwetrust/PL-Hydra-DDP-bug
$ cd PL-Hydra-DDP-bug && pip install -r requirements.txt
$ python bug_report_model.py
```

### Expected behavior

I would expect the program to use the subfolder spawned by the first process (rank 0) when loading the best checkpoint.

### Environment

```
* CUDA:
        - GPU:
                - GeForce GTX TITAN X
                - GeForce GTX TITAN X
        - available:         True
        - version:           10.2
* Packages:
        - numpy:             1.19.4
        - pyTorch_debug:     True
        - pyTorch_version:   1.7.0
        - pytorch-lightning: 1.0.5
        - tqdm:              4.54.1
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - 
        - processor:         x86_64
        - python:            3.7.9
        - version:           #219-Ubuntu SMP Tue Aug 11 12:26:50 UTC 2020
```

### Additional context

* For further details, please take a look at my recent [chat](https://hydra-framework.zulipchat.com/#narrow/stream/257092-Hydra-Torch/topic/Hydra.20.2B.20PL.20.2B.20DDP) with Hydra main author on Zulip.
* Take a look at this PL forums [topic](https://forums.pytorchlightning.ai/t/using-hydra-ddp/567/3).

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Fixes #5512.

If a user had their run dir set to `"data/${now:%Y-%m-%d_%H-%M-%S}"` when processes are spawned in the DDP accelerator, they would create new directories. This would lead to non-rank 0 processes referring to the wrong folder, and at test time they would crash with a file not found error. Annoyingly, they would also make new folders with the timestamp of spawn.

Fix this as suggested in #5512 by appending the hydra run dir before spawning the process. I've tested this locally and can confirm it works as expected!

Any ideas for a testing @tchaton? This is specific to the ddp accelerator might need to sync up on this.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/ddp_accelerator.py [hunk 1]
```diff
@@ -133,11 +133,13 @@ def _call_children_scripts(self):
...
 
             # start process
-            # if hydra is available and initialized, make sure to set the cwd correctly
+            # if hydra is available and initialized, make sure to set the original cwd correctly
+            # and pass current cwd for ddp processes (which hydra has overridden)
             cwd: Optional[str] = None
             if HYDRA_AVAILABLE:
                 if HydraConfig.initialized():
                     cwd = get_original_cwd()
+                    command += [f'hydra.run.dir={os.getcwd()}']
```
Comment (Borda) [inline pytorch_lightning/accelerators/ddp_accelerator.py:142]: this dir or rather the trainer default dir?

Comment (SeanNaren) [reply pytorch_lightning/accelerators/ddp_accelerator.py:142]: The Hydra directory I think is the right one to set, internally hydra has set this directory where the trainer directory is now inside (unless absolute I think), which fixes it as Omry suggested!

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Smart fix ! Can easily be tested with ddp and special test. We launch the test and make sure only 1 folder is being created.

File: pytorch_lightning/accelerators/ddp_accelerator.py (hunk 1 above)
Comment (omry) [inline pytorch_lightning/accelerators/ddp_accelerator.py:142]: You can use a different .hydra dir for each subjob so they don't overstep one another (or disable it completely.
currently, the subjobs will overwrite the files in the [.hydra](https://hydra.cc/docs/next/tutorials/basic/running_your_app/working_directory/) dir generated by the parent.

Disable:
```
hydra.output_subdir=null
```
Or use your local rank variable to give each subjob it's own .hydra dir:
```
hydra.output_subdir=.hydra_{local_rank}
```

By default, the log filename is derived from the job name. you can override that to something that depends on your local_rank to ensure that each subprocess gets their own job. (I didn't try it here but it should work):
```
hydra.job.name=subjob_{local_rank}
```

Comment (SeanNaren) [reply pytorch_lightning/accelerators/ddp_accelerator.py:142]: This is perfect, was my last worry! Will make this change and add a test

**Pull request merged**